### Part1. 计算手续费收入

1. 选定一个池子，获取池子的基本信息 ✅
2. 选定一个头寸和时间范围之后，获得这段时间内池子里的所有swap记录 ✅
3. 获取池子流动性的时间序列数据 ✅
4. 结合3和4计算出选定的头寸在这段时间内的手续费收入 

In [34]:
import pandas as pd
from datetime import datetime
from importlib import reload
from poolData import swapData
reload(swapData)

query = swapData.SwapDataQuery()

In [35]:
from datetime import datetime
begin = datetime(2022, 12, 31)
end = datetime(2023, 1, 1)
pool_id = "0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640"

### 1. 获取池子的基本信息

https://etherscan.io/address/0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640#readContract

手续费率：0.05%

protocol fee: 0.05%

### 2. 获取所有swpa记录

In [36]:
data = query.query_swaps(begin=int(begin.timestamp()), end=int(end.timestamp()), pool_id=pool_id)
data.head()

this is the cursor: 0x2e419db5f7dbc176ce32523a544edd34c2104d897377ceafcd70e98f7fab1b26#3955300
this is the cursor: 0x5bb913f1604bc2b8ead2c12020bbec14ba575fcbd157338be8a52b1b383b550d#3957295
this is the cursor: 0x88636eddfd98e3c09e82a6debf5d70582815381069ddf252ba7e8ee5595221ae#3953732
this is the cursor: 0xb69f6ce3a688cf74bdc290520146a400c7800d7271e0b76382a2eb8414ac7ccc#3957867
this is the cursor: 0xe52d316a15879501810d0a1bb4d7cb680f1d6c5350474d82467754d80346baa1#3954249
this is the cursor: 0xffc998053eff81239ee9aed6a27ee915799402b02b45d5df56705f136425bbb0#3958244


id   timestamp  \
0  0x0002088b2f326f0114b41b8a89d7f15f12d293328a6a...  1672447895   
1  0x0013b93196471c4dcb508221cd6a0bbda1e6bdc654b3...  1672421291   
2  0x0013fce47ecd63a38700ff544d31d9cbf95ccdfccb02...  1672422803   
3  0x00151559f96c8ad818a4847f123c1f14fa4974f84b09...  1672438979   
4  0x001728914057df211c141e2183ba0409d30c1c099ff4...  1672436315   

                                       sender  \
0  0x53222470cdcfb8081c0e3a50fd106f0d69e63f20   
1  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
2  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
3  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
4  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   

                                    recipient  \
0  0x53222470cdcfb8081c0e3a50fd106f0d69e63f20   
1  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
2  0x0ddd2d26db56de1cc755b56dd3bca269dd6d8955   
3  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
4  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   

                                       origin       amount0  \
0  0xff57ae5870529b80ea01d403c3f8bd94d11b9678   4266.117177   
1  0x899c059a72694e16608a14aae420022eb55cdefb          1000   
2  0x0ddd2d26db56de1cc755b56dd3bca269dd6d8955  -5011.751045   
3  0x9902d3a02f0c35a85095de584ace59bb3538f6aa   2824.243528   
4  0x87e485cb631d7616256076a9cb64fff5c7baa468     99.563794   

                 amount1                            amountUSD  \
0  -3.560090738149703786  4264.941859937119356472465091232485   
1  -0.836888527815654406  999.7670959814923754131727434898885   
2                    4.2   5014.00285661645623131938387967046   
3  -2.355025735643548311  2822.308348680553578867706064338535   
4  -0.082988052797955076  99.47429731086213279113076465711465   

                         sqrtPriceX96    tick logIndex  
0  2289294668390795750019773769053102  205438       13  
1  2292565998538329889197212915778332  205467      259  
2  2292987089929314170924596543568948  205470       81  
3  2288414245600305261854951574027895  205431       48  
4  2287943199586960584738206920087911  205426      322

### 3. 获取池子流动性的时间序列数据

In [37]:
data = query.query_liquidity(begin=int(begin.timestamp()), end=int(end.timestamp()), pool_id=pool_id)
data.head()

this is the cursor: 1672498800


periodStartUnix             liquidity                           sqrtPrice  \
0       1672419600  43383970859571409238  2292741488478236193954710948079555   
1       1672423200  41908297001142436164  2289914226033824321765933548671387   
2       1672426800  43259411757668858895  2292095502081045216146554614279866   
3       1672430400  41853986357057955972  2290601372199165019827373276558852   
4       1672434000  59109260845469698637  2288829923181971195179725662338341   

                           token0Price  \
0  1194.122792146251869705949121256262   
1  1197.073280553630480706658776824827   
2  1194.795971584433226686676411440843   
3  1196.355180105557782621625188956052   
4  1198.207744422359329581249063675693   

                               token1Price    tick  \
0  0.0008374348153950348453709271537854136  205468   
1   0.000835370746507275888776747615886998  205444   
2  0.0008369629826202778712332352962748905  205463   
3  0.0008358721695941226979286513996652267  205450   
4  0.0008345798169432523948316314255772096  205434   

                 feeGrowthGlobal0X128  \
0  2186941498930609995212528368989333   
1  2186952021398667747348751967478818   
2  2186954913044984967466704176819519   
3  2186962943960877559886609212538257   
4  2186981817612323087894790173617207   

                          feeGrowthGlobal1X128  \
0  1031090019324772402022075697810397089878707   
1  1031092959063940135596058161042444273527573   
2  1031099918456639174812780045987174854968176   
3  1031103373243662410724723254463120782143002   
4  1031115295566281377942677265799122428836270   

                                tvlUSD volumeToken0 volumeToken1 volumeUSD  \
0  333609426.6969159360891818260545127            0            0         0   
1  333488757.0211926141408956767590597            0            0         0   
2   333637095.322414727186961463831933            0            0         0   
3  333906894.4645702736186860023230338            0            0         0   
4  334440380.7305554361225720897040363            0            0         0   

  feesUSD txCount                                 open  \
0       0     225  1194.511809050204970771435064936336   
1       0     260  1194.123444963926810270341103316625   
2       0     289  1197.073280553630480706658776824827   
3       0     263  1194.796131311705043167380267771029   
4       0     338  1196.394407853511805846627602601061   

                                  high                                  low  \
0  1195.135863873698740398661399390146  1193.804391237179988357396971990688   
1  1197.073280553630480706658776824827  1194.080786262999696020970218231062   
2   1197.07382117549381742907326713768  1194.750385533700741374110953471534   
3  1196.355180105557782621625188956052  1194.693649299187396892654724716834   
4  1200.916144596237448207735051238417  1196.394407853511805846627602601061   

                                 close  
0  1194.122792146251869705949121256262  
1  1197.073280553630480706658776824827  
2  1194.795971584433226686676411440843  
3  1196.355180105557782621625188956052  
4  1198.207744422359329581249063675693

### 4. 计算头寸的手续费分成收入

参数解释：
1. 不要忘记protocol fee(需要从池子合约中获取)

    swap的手续费不是所有都给流动性提供者的，一部分会给协议，最终由dao收取。

2. feeGrowthGlobal0X128 和 feeGrowthGlobal1X128

    feeGrowthGlobal0X128:  The all-time global fee growth, per unit of liquidity, in token0

    feeGrowthGlobal1X128:  The all-time global fee growth, per unit of liquidity, in token1

    这是池子合约中的两个全局状态变量，池子每一次swap都会自动更新这两个值，表示的是整个池子里手续费的累计增长率，X128表示这个值是乘$ 2 ^ {128} $之后的值，所以在实际使用的时候需要把这个值除以$ 2 ^ {128} $.

    feeGrowthGlobal的计算方法：

    $$ feeGrowthGlobal0X128 = \frac{accumulative fees \times 2 ^ {128}}{total liquidity} $$

3. feeGrowthOutside0X128 和 feeGrowthOutside1X128

    这两个变量的定义和上面的global差不多，差别是feeGrowthOutside0X128是记录一个tick外所有的手续费增长率

    知道了feeGrowthOutside0X128和feeGrowthGlobal0X128的值，就可以计算出每次swap之后该position能获得的手续费是多少。
    
考虑两种情况：

a. swap是在你的tick range内发生的

$$ feetoken0 = \frac {(feeGrowthGlobal0X128 - feeGrowthOutside0X128_lower - feeGrowthOutside0X128_upper - feeGrowthInside0LastX128) \times liquidity}{ 2 ^ {128} \times 10 ^ {decimal}} $$

b. swap发生后你的tick out of range了

    i. current ticker > tick upper
        $$ token0fee = \frac {(feeGrowthOutside0X128_upper - feeGrowthOutside0X128_lower - feeGrowthInside0LastX128) \ times liquidity } { 2 ^ {128}  \times 10 ^ {decimal} } $$
    
    ii. current ticker < tick upper
        $$ token0fee = \frac {(feeGrowthOutside0X128_lower - feeGrowthOutside0X128_upper - feeGrowthInside0LastX128) \ times liquidity } { 2 ^ {128}  \times 10 ^ {decimal} } $$

***注意***

在the graph中能拿到的最细粒度数据是小时级别的，用以上的计算方法存在一个误差点：

利用小时级别的feeGrowth数据来计算手续费收入，忽视了一个小时内池子流动性发生的变化。如果在这一个小时内，有流动性提供者推出或者新的流动性提供者加入，那么feeGrowth数据其实是会发生变化的。

但这个误差是可以容忍的，所以我们假设一个小时内池子的流动性不发生变化。

4. fee

在v2中，收取的手续费会自动转变成liquidity，随着池子内swap不断发生，你的liquidity也会逐渐积累。

但在v3中，收取的手续费不再自动转成liquidity，背后的原因是：virtual reserve的存在导致收取的手续费无法直接按比例加入池子成为流动性。

fee的思考：

v3的特点就是允许你在某几个tick内提供流动性，需要一个比较好的方式来存储以tick为单位的手续费，所以v3记录手续费的方式是用全局变量记录整个池子内的手续费增长情况 + 每个tick的手续费增长情况，和直接记录手续费大小相比这是一个优化后的数据结构设计。这一部分的内容在白皮书的6.3左右：

swap的流程：
![swap-control](formulas/swap-control.png)

对于每一个tick，合约中都会存储以下这些变量：

1. liquidityNet：

    the total amount of liquidity that should be kicked in or out when the tick is crossed. 
    <!-- $$ \Delta L = \frac {} $$ -->
2. liquidityGross
3. feeGrowthOutside0X128
4. feeGrowthOutside1X128
5. secondsOutside
6. tickCumulativeOutside
7. secondsPerLiquidityOutsideX128

另外这个文档中也有很详细的解释：
https://uniswapv3book.com/docs/milestone_5/swap-fees/#:~:text=Fees%20are%20paid%20by%20users,amount%20divided%20by%20pool's%20liquidity).

To make fees accounting simpler, Uniswap V3 tracks the global fees generated by 1 unit of liquidity. Price range fees are then calculated based on the global ones: fees accumulated outside of a price range are subtracted from the global fees. Fees accumulated outside of a price range are tracked when a tick is crossed (and ticks are crossed when swaps move the price; fees are collected during swaps). With this approach, we don’t need to update fees accumulated by each position on every swap–this allows to save a lot of gas and make interaction with pools cheaper.

### 5. 获取tick的数据

获取一个池子内所有的被initialized的ticks

In [38]:
data = query.query_ticks(poolAddress="0x8ad599c3a0ff1de082011efddc58f1908eb6e6d8")
data

tickIdx   liquidityGross     liquidityNet  \
0        0                0                0   
1   105960                0                0   
2    -1080  955255939572502  955255939572502   
3   115140                0                0   
4   120660                0                0   
..     ...              ...              ...   
95  173400                0                0   
96  173700   16976441980382   16976441980382   
97  173820                0                0   
98   -1740                0                0   
99  174180                0                0   

                                  price0  \
0                                      1   
1    39953.44719942822385402034558214009   
2   0.8976324433094195884383843668365469   
3    100049.8997864926770132351218727111   
4    173754.1660171966344141422511922068   
..                                   ...   
95   33907034.48633657555701262300435812   
96   34939605.01341065627733928620852156   
97   35361384.80257631179772826203761658   
98  0.8403042077859023045607653103039314   
99   36657520.31640865803259235355450485   

                                         price1 volumeToken0 volumeToken1  \
0                                             1            0            0   
1      0.00002502912940173810732272813416956387            0            0   
2           1.114041729945910239807309095780991            0            0   
3     0.000009995012510097544967562707071231679            0            0   
4     0.000005755257689194220169430331041597795            0            0   
..                                          ...          ...          ...   
95  0.00000002949240519405987126391149071657875            0            0   
96  0.00000002862081582250789831846447283671908            0            0   
97  0.00000002827943547977632812446748555012594            0            0   
98          1.190045213072152005202431109200761            0            0   
99  0.00000002727953204058866708059449100626584            0            0   

   volumeUSD untrackedVolumeUSD feesUSD collectedFeesToken0  \
0          0                  0       0                   0   
1          0                  0       0                   0   
2          0                  0       0                   0   
3          0                  0       0                   0   
4          0                  0       0                   0   
..       ...                ...     ...                 ...   
95         0                  0       0                   0   
96         0                  0       0                   0   
97         0                  0       0                   0   
98         0                  0       0                   0   
99         0                  0       0                   0   

   collectedFeesToken1 collectedFeesUSD createdAtTimestamp  \
0                    0                0         1622488494   
1                    0                0         1623967517   
2                    0                0         1620937986   
3                    0                0         1622536811   
4                    0                0         1623419807   
..                 ...              ...                ...   
95                   0                0         1620593461   
96                   0                0         1621115540   
97                   0                0         1623664594   
98                   0                0         1659200074   
99                   0                0         1622166087   

   liquidityProviderCount              feeGrowthOutside0X128  \
0                       0                                  0   
1                       0                                  0   
2                       0  119690017018605194611917554219251   
3                       0                                  0   
4                       0                                  0   
..                    ...                                ...   
95                    

## 关于回测框架

计算头寸的收益最难的就是手续费的部分了，但是如果有position feeGrowthOutside1X128的时间序列数据，结合池子的feeGrowthGlobal0X128对时间序列数据，可以直接计算出一个头寸的手续费收入。

不过为了分析不同策略的pnl（如果不自己写一个的话，只能用现实中真实发生过的头寸进行分析，没有办法做一些策略的设计），我们需要一个回测框架，支持模拟各种头寸的设置，假设头寸不影响池子的深度和价格的情况下分析头寸的pnl。

那么如何计算手续费呢？

两种方法：

### 方法一、
1. 获取池子内所有的swap数据，筛选出在你的头寸之内的所有swap

2. 计算该swap会给你的头寸带来多少手续费收入

### 方法二、
用合约中记录的状态变量：global和outside fee growth

但这个方法只能研究现有的头寸，除非：

**但是我突然有个大胆的想法🙋🏽**

uniswap中的池子不同于传统的金融市场，传统金融市场中做回测是很简单的。比如你想做一个股票的回测框架，你可以不用考虑你的交易对市场价格产生的影响，因为上市的股票往往有着足够的深度和流动性。但是在uniswap中，许多池子都非常非常小，你的头寸很可能会大幅度影响币对的价格。

如果只研究池子深度足够的池子里什么策略好，那上面所说的回测框架是足够了的，但是研究出来的结果对于深度小的池子肯定是不适用的。

能不能自己部署一个uniswap合约，交易数据可以用uniswap上的实际交易数据或者用中心化交易所的交易数据，把这些数据放到自己部署的合约上重放。

· 选一个很深的池子
· 选很小的头寸（这样的话交易gas fee会让pnl为负，也不太好
    ·消除手续费的影响？用在池子内提供mock的token0和token1作为策略的base line？参考：https://pub.tik.ee.ethz.ch/students/2021-HS/BA-2021-21.pdf
    

### 计算手续费分成收入（方法一）

In [39]:
# 先拿一个池子，随便选一个头寸，以便拿计算出来的手续费和网站上的真实手续费比较
data = query.query_positions(pool_id=pool_id, block_gte=16859000, limit=10, orderBy="owner")
data

pool_id: 0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640 - type: <class 'str'>


id                                       owner liquidity  \
0  414897  0x0000000000000000000000000000000000000000         0   
1  474569  0x0000000000000000000000000000000000000000         0   
2  474602  0x0000000000000000000000000000000000000000         0   
3  475738  0x0000000000000000000000000000000000000000         0   
4  477339  0x0000000000000000000000000000000000000000         0   
5  477344  0x0000000000000000000000000000000000000000         0   
6  477361  0x0000000000000000000000000000000000000000         0   
7  477373  0x0000000000000000000000000000000000000000         0   
8  477408  0x0000000000000000000000000000000000000000         0   
9  477430  0x0000000000000000000000000000000000000000         0   

  depositedToken0        depositedToken1 withdrawnToken0  \
0     8470.656242   1.579300632349748811     9646.002942   
1     1546.194712                      0     1546.194711   
2         751.276   0.430281298965067752      758.024421   
3    44265.292808  21.992013333106751863    46155.041039   
4     7207.835608   1.624088891207906744     6789.330541   
5     5127.178576   2.844458176701934854               0   
6      6555.14594   2.014347420598452127     6794.317657   
7     3006.479542   4.088212014049604052    10004.892651   
8     5594.224577   2.569014555414213226               0   
9     4360.945466   3.302134333461079045     4520.753418   

         withdrawnToken1 collectedFeesToken0 collectedFeesToken1  \
0   0.880907147308836156                   0                   0   
1                      0                   0                   0   
2    0.42652441287500024                   0                   0   
3  20.937462667557817877                   0                   0   
4   1.868783764506809232                   0                   0   
5   5.838929534189147607                   0                   0   
6   1.874497253498284477                   0                   0   
7                      0                   0                   0   
8   5.832840993957891698                   0                   0   
9   3.208555011603192752                   0                   0   

                            feeGrowthInside0LastX128  \
0                  254388150839429989634847833044971   
1                  182646244709656214368682468292997   
2                  165030886172350255418678531400564   
3                 2327263699268187470644153361387606   
4                   38174011643217368409761149577587   
5  1157920892373161954235709850086879078532699845...   
6                   38174011643217368409761149577587   
7  1157920892373161954235709850086879078532699845...   
8                  424678637577946270863037481337319   
9  1157920892373161954235709850086879078532699846...   

                            feeGrowthInside1LastX128  \
0         130416506739465878095744363051182666646616   
1         119528223954720344066558179677182575381622   
2          91997372302310571062261748791992375307046   
3        1118077605153562457755372148609081657595811   
4          30676715703607205933899356875892515876396   
5  1157920892373161954235709850086879077458955759...   
6          30676717780587031755586142063758423585118   
7  1157920892373161954235709850086879077852150618...   
8         297853581244479983528068081759041114737915   
9  1157920892373161954235709850086879078452521653...   

                                    token0.id  \
0  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
1  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
2  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
3  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
4  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
5  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
6  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
7  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
8  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   
9  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   

                                    token1.id  \
0  0xc02aaa39b223fe8d0a0e5c4f27ead90

In [69]:
# 选定池子,头寸,时间：
pool_id = "0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640"
position_id = 469767

'''
注意1:  poolid在uniswap中是每个头寸的唯一标识:
positionId = keccak256(abi.encodePacked(
    bytes32(poolAddress),
    bytes32(liquidityProvider),
    int24(lowerTick),
    int24(upperTick),
    uint24(fee)
))

注意2: 算完之后可以参考这个看算的对不对：
https://revert.finance/#/uniswap-position/mainnet/469767
'''

data = query.query_position(position_id=position_id)
data


id                                       owner        liquidity  \
0  469767  0x56209a886c507f8725c822b2b1f77ef4c9dc853c  286206700622530   

  depositedToken0       depositedToken1 withdrawnToken0 withdrawnToken1  \
0      1466.04493  0.865982013077301422               0               0   

  collectedFeesToken0 collectedFeesToken1           feeGrowthInside0LastX128  \
0                   0                   0  907373610974648595634682261671971   

                     feeGrowthInside1LastX128  \
0  578710287513428657835512647395860822798555   

                                    token0.id  \
0  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   

                                    token1.id  \
0  0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2   

                                        tickLower.id  \
0  0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640#198760   

                                        tickUpper.id  \
0  0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640#204080   

                                      transaction.id  
0  0xe2e2041d2b09d5d7ae4f999951775f7ad527a64f3ad1...

In [70]:
data["transaction.id"][0]

'0xe2e2041d2b09d5d7ae4f999951775f7ad527a64f3ad103584bfdb34307056a9d'

### 先简单看一下这个头寸的基本信息：
1. 池子: USDC/WETH
    token0: usdc
    token1: weth
    价格 p = y/x = p_weth / p_usdc
2. usdc: 1466.04493
   eth: 0.865982013077301422
3. 创建时间: March 19, 2023 at 2:17 PM tx：https://etherscan.io/tx/0xe2e2041d2b09d5d7ae4f999951775f7ad527a64f3ad103584bfdb34307056a9d
4. 价格区间：
    需要根据ticker计算一下

$$p(i) = 1.0001^i$$


另外在token合约中存在decimal，usdc的decimal是6，weth的decimal是18。而uniswap在计算的时候是不管decimal的，所以他的价格实际上是：amount_usdc * （10 ** -6）/ amount_weth * (10 ** -18)，整理得到：

$$ p_real = p * (10 ** -12)

In [41]:
p_lower = (1.0001 ** 198760) * (10 ** -12)
p_upper = (1.0001 ** 204080) * (10 ** -12)
print("===price(y/x) the usdc price in terms of weth===")
print("lower price is:", p_lower)
print("higher price is:", p_upper)
p_lower = 1 / ((1.0001 ** 198760) * (10 ** -12))
p_upper = 1 / ((1.0001 ** 204080) * (10 ** -12))
print("===It's easier to read weth price in terms of usdc, so we might inverse the numbers: price(x/y) the weth price in terms of usdc===")
print("lower price is:", p_lower)
print("higher price is:", p_upper)

===price(y/x) the usdc price in terms of weth===
lower price is: 0.00042815946513519496
higher price is: 0.0007288508458564415
===It's easier to read weth price in terms of usdc, so we might inverse the numbers: price(x/y) the weth price in terms of usdc===
lower price is: 2335.5784034442436
higher price is: 1372.0228297532435


In [42]:
# 选定时间
begin = datetime(2023, 3, 19)
end = datetime.now()
# 获取该时段内池子里所有的swap
swaps = query.query_swaps(begin=int(begin.timestamp()), end=int(end.timestamp()), pool_id=pool_id)
swaps.head()

this is the cursor: 0x02a679e24a73775694639ae501a83ce9dda883369f01a6e0ab8083d43f27f80a#4711857
this is the cursor: 0x0558dac2cef364cc3e2e19071953e1ee0982d401fd0189692e4854451c3c1109#4681779
this is the cursor: 0x0831602934625f966c9f52574a273f332d2d2fd5f0d60974caedfcc80982c7e1#4704899
this is the cursor: 0x0ade474e0488eb09852e77e5cd97d7095fd3639b18ae51bf51c33eed4ddfd162#4660089
this is the cursor: 0x0dab39e827e8a05047cfe31b4e379ad7ef1515c7c6fa9adab65b8e1de8ee4afe#4692566
this is the cursor: 0x1099121a5c6bea603c5630109e68e5558758d1c475f76890b0c44ebdb7c27374#4645270
this is the cursor: 0x135de5bcb320ce42cdbef409e5ef920ddc8586e15f8c94a6e4446b66446a1456#4704029
this is the cursor: 0x162721acfdb0fb1c6c0bdfdfe39ec50600f0e2190ffcff86213d0e4fbf9fdcac#4627183
this is the cursor: 0x18f0f150d451f653b2e80bf7a3c4faa07213853416f83677256d733b65bfd48f#4685016
this is the cursor: 0x1bbc921864717db712271be0072319c2feb86ce785fbb2d88a36434f6542dd3a#4652552
this is the cursor: 0x1e6ce1f443be2fb61c960f130da4

id                                       owner        liquidity  \
0  469767  0x56209a886c507f8725c822b2b1f77ef4c9dc853c  286206700622530   

  depositedToken0       depositedToken1 withdrawnToken0 withdrawnToken1  \
0      1466.04493  0.865982013077301422               0               0   

  collectedFeesToken0 collectedFeesToken1           feeGrowthInside0LastX128  \
0                   0                   0  907373610974648595634682261671971   

                     feeGrowthInside1LastX128  \
0  578710287513428657835512647395860822798555   

                                    token0.id  \
0  0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48   

                                    token1.id  \
0  0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2   

                                        tickLower.id  \
0  0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640#198760   

                                        tickUpper.id  \
0  0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640#204080   

                                      transaction.id  
0  0xe2e2041d2b09d5d7ae4f999951775f7ad527a64f3ad1...

### 参数解释
首先要理解swap表中的参数。 首先这个表是通过监听uniswap合约中的swap事件得到的，关于swap事件：https://docs.uniswap.org/contracts/v3/reference/core/interfaces/pool/IUniswapV3PoolEvents#swap

#### sqrtPriceX96：
The sqrt(price) of the pool after the swap, as a Q64.96

存储的方式是X96，所以得处理一下，除以2**96，得到sqrt p，然后平方一下得到p

$$ sqrtPriceX96  = \sqrt{p} \times 2 ^ {96}$$

#### liquidity：
The liquidity of the pool after the swap
#### tick：
The log base 1.0001 of price of the pool after the swap


In [44]:
swaps

id   timestamp  \
0      0x00005be3feffa40600371196f6ebe99484cef2e030cb...  1679680919   
1      0x00005e650a5a29a7340961ee1070bcb4d699a5c6b47c...  1680218951   
2      0x00010099f534a61a99fa906e57cabd736e9f7e43ee0f...  1679774651   
3      0x000128f2e478320e005d9435b14ad71341d06b8d1d7f...  1679404235   
4      0x0001770e8cba1be004b066570f49bf727e1b00774d32...  1680084887   
...                                                  ...         ...   
92317  0xfffcffa89359b3b5e69363eae9928ac04f9e82d19704...  1679299679   
92318  0xfffdd7e583a0cec9a6427430f76f6d8cf7af6aa4ff97...  1679319839   
92319  0xffff6cefeb9b19bfb66b1f036307a8bef5f6055cbf0d...  1679611667   
92320  0xffff7c08cb22ae1f06c6792e49a67169faeba6202ecd...  1679476367   
92321  0xffffbfa55e57618b47f076e56b7579e64bb87996b402...  1679243279   

                                           sender  \
0      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
1      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
3      0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
4      0x7122db0ebe4eb9b434a9f2ffe6760bc03bfbd0e0   
...                                           ...   
92317  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
92318  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
92319  0xe8cfad4c75a5e1caf939fd80afcf837dde340a69   
92320  0x0c7847778e9025991fc1e90e6caef4011435decb   
92321  0x1111111254eeb25477b68fb85ed929f73a960582   

                                        recipient  \
0      0x697d1eeb266aec5251901e20e6a09dbdb68008c1   
1      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2      0xd22125cb225f39fe6aa0c16afe6b75872a3529d1   
3      0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5   
4      0x1111111254eeb25477b68fb85ed929f73a960582   
...                                           ...   
92317  0xca311ff7afe460732ab8175f1cf06a1006e4741d   
92318  0xd918252c46bf5d399ce827151b422810388c79ec   
92319  0xe8cfad4c75a5e1caf939fd80afcf837dde340a69   
92320  0x0c7847778e9025991fc1e90e6caef4011435decb   
92321  0x1111111254eeb25477b68fb85ed929f73a960582   

                                           origin         amount0  \
0      0x697d1eeb266aec5251901e20e6a09dbdb68008c1   -19922.526611   
1      0xc0c441bd9889f30fd219478a985e84fac8c2a706     1104.872406   
2      0xd22125cb225f39fe6aa0c16afe6b75872a3529d1   -12033.180287   
3      0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5    -5000.197259   
4      0xc6c7565644ea1893ad29182f7b6961aab7edfed0     -794.907298   
...                                           ...             ...   
92317  0xca311ff7afe460732ab8175f1cf06a1006e4741d             817   
92318  0xd918252c46bf5d399ce827151b422810388c79ec    -2057.963808   
92319  0x24f7ef98522dd61d529464f67bb3ffe96ea8afc2  -489300.023033   
92320  0xb720c31d4cf1c730f21f6e59aa20fc65bdb7acf6      -95.891668   
92321  0x610bb423d080dbbb9ea5fb8bab100aad2d81bca4           10000   

                      amount1                            amountUSD  \
0       11.442382101497925129  19903.00432076901806488783284947735   
1       -0.617085078176457911  1103.966933243716954948609702966833   
2                           7  12040.45320764383144992175201712345   
3        2.750357643401229741  4998.420143808785696912599946213698   
4        0.438164882921890842   795.916771734336092480722470468768   
...                       ...                                  ...   
92317   -0.452283223933597064  817.4212013095482548812137296315135   
92318    1.168116903666099567  2059.262184665299100172312721860874   
92319  270.321599613403480029  489606.2536095093340911207546503689   
92320    0.053563039734003172  95.98812474424707318087003399598465   
92321   -5.535830737955111222  9997.574446181269182813047135046025   

                             sqrtPriceX96    tick logIndex  
0      1898284365247915042890667106569354  201692      295  
1      1872856442624146468221690277382731  201422      369  
2      1910435929075656551647679638334639  201820      272  
3

In [45]:
# convert str into float
swaps['amount0'] = swaps['amount0'].astype(float)
swaps['amount1'] = swaps['amount1'].astype(float)
swaps['amountUSD'] = swaps['amountUSD'].astype(float)
swaps['sqrtPriceX96'] = swaps['sqrtPriceX96'].astype(float)
swaps['tick'] = swaps['tick'].astype(int)
swaps['logIndex'] = swaps['logIndex'].astype(int)
# transform X86 into normal number
swaps["sqrtPriceX96"] = (swaps["sqrtPriceX96"] / (2**96) ) ** 2
# transform sqrt_p into p
swaps["sqrtPriceX96"] = 1 / (swaps["sqrtPriceX96"] * (10 ** -12))


In [46]:
swaps

id   timestamp  \
0      0x00005be3feffa40600371196f6ebe99484cef2e030cb...  1679680919   
1      0x00005e650a5a29a7340961ee1070bcb4d699a5c6b47c...  1680218951   
2      0x00010099f534a61a99fa906e57cabd736e9f7e43ee0f...  1679774651   
3      0x000128f2e478320e005d9435b14ad71341d06b8d1d7f...  1679404235   
4      0x0001770e8cba1be004b066570f49bf727e1b00774d32...  1680084887   
...                                                  ...         ...   
92317  0xfffcffa89359b3b5e69363eae9928ac04f9e82d19704...  1679299679   
92318  0xfffdd7e583a0cec9a6427430f76f6d8cf7af6aa4ff97...  1679319839   
92319  0xffff6cefeb9b19bfb66b1f036307a8bef5f6055cbf0d...  1679611667   
92320  0xffff7c08cb22ae1f06c6792e49a67169faeba6202ecd...  1679476367   
92321  0xffffbfa55e57618b47f076e56b7579e64bb87996b402...  1679243279   

                                           sender  \
0      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
1      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
3      0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
4      0x7122db0ebe4eb9b434a9f2ffe6760bc03bfbd0e0   
...                                           ...   
92317  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
92318  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
92319  0xe8cfad4c75a5e1caf939fd80afcf837dde340a69   
92320  0x0c7847778e9025991fc1e90e6caef4011435decb   
92321  0x1111111254eeb25477b68fb85ed929f73a960582   

                                        recipient  \
0      0x697d1eeb266aec5251901e20e6a09dbdb68008c1   
1      0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2      0xd22125cb225f39fe6aa0c16afe6b75872a3529d1   
3      0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5   
4      0x1111111254eeb25477b68fb85ed929f73a960582   
...                                           ...   
92317  0xca311ff7afe460732ab8175f1cf06a1006e4741d   
92318  0xd918252c46bf5d399ce827151b422810388c79ec   
92319  0xe8cfad4c75a5e1caf939fd80afcf837dde340a69   
92320  0x0c7847778e9025991fc1e90e6caef4011435decb   
92321  0x1111111254eeb25477b68fb85ed929f73a960582   

                                           origin        amount0     amount1  \
0      0x697d1eeb266aec5251901e20e6a09dbdb68008c1  -19922.526611   11.442382   
1      0xc0c441bd9889f30fd219478a985e84fac8c2a706    1104.872406   -0.617085   
2      0xd22125cb225f39fe6aa0c16afe6b75872a3529d1  -12033.180287    7.000000   
3      0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5   -5000.197259    2.750358   
4      0xc6c7565644ea1893ad29182f7b6961aab7edfed0    -794.907298    0.438165   
...                                           ...            ...         ...   
92317  0xca311ff7afe460732ab8175f1cf06a1006e4741d     817.000000   -0.452283   
92318  0xd918252c46bf5d399ce827151b422810388c79ec   -2057.963808    1.168117   
92319  0x24f7ef98522dd61d529464f67bb3ffe96ea8afc2 -489300.023033  270.321600   
92320  0xb720c31d4cf1c730f21f6e59aa20fc65bdb7acf6     -95.891668    0.053563   
92321  0x610bb423d080dbbb9ea5fb8bab100aad2d81bca4   10000.000000   -5.535831   

           amountUSD sqrtPriceX96    tick  logIndex  
0       19903.004321  1741.953774  201692       295  
1        1103.966933  1789.576173  201422       369  
2       12040.453208  1719.864424  201820       272  
3        4998.420144  1818.915412  201260       209  
4         795.916772  1815.080405  201281        11  
...              ...          ...     ...       ...  
92317     817.421201  1805.489555  201334       184  
92318    2059.262185  1762.656835  201574       292  
92319  489606.253610  1810.431169  201307         2  
92320      95.988125  1791.153569  201414        57  
92321    9997.574446  1805.539466  201334       222  

[92322 rows x 11 columns]

In [47]:
# 首先不考虑tick被击穿的情况, 表中的tick是本次swap之后池子内的swap，为了先简化计算，假设最终的tick在我们选定的头寸之内的所有swaps都是在我们的头寸之内发生的
my_tick_lower = 198760
my_tick_upper = 204080
# 筛选swaps的条件：swap后最终的tick在我们的头寸tick之内
condition = (swaps['tick'] > my_tick_lower) & (swaps['tick'] < my_tick_upper)
# 筛选
print("total swaps: \n", swaps.shape[0])
swaps = swaps[condition]
print("total swaps in range: \n", swaps.shape[0])
# 好吧这个人的range选的太太大了，一直in range, 不过这样算手续费就很准

total swaps: 
 92322
total swaps in range: 
 92322


In [48]:
# 来计算手续费
# 先复制一个swap 用来乱算
swaps1 = swaps
# 先把不用的列去掉
swaps1 = swaps1.drop(swaps1.columns[[0, 2, 3, 4, 7, 10]], axis = 1)
swaps1.head(6)

timestamp        amount0     amount1 sqrtPriceX96    tick
0  1679680919  -19922.526611   11.442382  1741.953774  201692
1  1680218951    1104.872406   -0.617085  1789.576173  201422
2  1679774651  -12033.180287    7.000000  1719.864424  201820
3  1679404235   -5000.197259    2.750358  1818.915412  201260
4  1680084887    -794.907298    0.438165  1815.080405  201281
5  1679879915 -226399.210721  127.909247  1770.542895  201529

#### position liquidity
计算手续费的时候还有一个重要的点是头寸的流动性，因为手续费是按照头寸流动性占总流动性的比例派发的。

v2中计算你的liquidity是多少非常简单：

$$ L = x \times y $$

在v3中由于virtual reserve的存在，计算liquidity的公式变成（推导参考：https://uniswapv3book.com/docs/milestone_1/calculating-liquidity/#liquidity-amount-calculation和https://atiselsts.github.io/pdfs/uniswap-v3-liquidity-math.pdf

推导过程：

首先在v3中流动性的式子变成了这样：

$$ (x_{real} + \frac {L} {\sqrt{p_{b}}}) \times (y_{real} +  L \times {\sqrt{p_{a}}}) = L ^2 $$
(pa是头寸的价格下限，pb是上限)

当$ P \le p_{a} $的时候，池子里只有x，y = 0

$$ (x_{real} + \frac {L} {\sqrt{p_{b}}}) \times  L \times {\sqrt{p_{a}}} = L ^2 $$

当$ P \ge p_{b} $的时候，池子里只有y，x = 0

$$ \frac {L} {\sqrt{p_{b}}} \times (y_{real} +  L \times {\sqrt{p_{a}}}) = L ^2 $$

整理上面的两个式子，可以得到L的两个计算公式：

1. 

$$ L = x \times \frac {\sqrt{p_{a}} \cdot \sqrt{p_{b}}} {\sqrt{p_{a}} - \sqrt{p_{b}}} $$

2. 

$$ L = \frac {y} {\sqrt{p_{a}} - \sqrt{p_{b}}} $$

将两个值都计算出来之后，选择小的那一个。

### 计算tick range内的总流动性

In [49]:
# $$ sqrtPriceX96  = \sqrt{p} \times 2 ^ {96}$$
p_lower = (1.0001 ** 198760) 
p_upper = (1.0001 ** 204080) 

sqrt_p_lower = (p_lower ** 0.5) * (2 ** 96)
sqrt_p_upper = (p_upper ** 0.5) * (2 ** 96)

print('this is sqrt price lower: {:.0f}'.format(sqrt_p_lower))
print('this is sqrt price upper:{:.0f}'.format(sqrt_p_upper))



this is sqrt price lower: 1639390289596216154630867994542080
this is sqrt price upper:2138941539491792332891588793466880


In [50]:
# 首先我们已经有了池子liquidity所有tick的流动性数据，需要计算出池子里在我们的头寸范围内的总流动性
# 我本来以为算这个会很麻烦，但是我发现query_liquidity函数返回的数据中，liquidity就是in range的liquidity。
liquidity_data = query.query_liquidity(begin=1679068800, end=1679673600, pool_id="0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640", sqrtPrice_gt="1639390289596216154630867994542080", sqrtPrice_lt="2138941539491792332891588793466880")
liquidity_data

this is the cursor: 1679670000


periodStartUnix              liquidity  \
0         1679072400   30788832203141306590   
1         1679076000   30335451111637185983   
2         1679079600   29136272383745257041   
3         1679083200   29155526732098504706   
4         1679086800   27547634722926741243   
..               ...                    ...   
162       1679655600  366641280831821063789   
163       1679659200   28728899486496684877   
164       1679662800   17341134938965724730   
165       1679666400   25891731533131139723   
166       1679670000   26624256669432733268   

                              sqrtPrice                          token0Price  \
0    1906022337482011155206725434245622  1727.838693579332397287823785551691   
1    1902228344319510065920412019047373  1734.737913210422806755592553486249   
2    1896207166369707888540506256730392  1745.772308123954713225775051712259   
3    1897039698024161669848304803585763  1744.240350950043333371334764406051   
4    1887391159854370169541208055025004  1762.119407182205743461631236756587   
..                                  ...                                  ...   
162  1890497417509973863512413999058108   1756.33352242884865230977759799148   
163  1880529834690179903284307439954389  1775.001447909970790514255477107159   
164  1883883812814254233295026104455764  1768.686815888479568803384012848543   
165  1884986754053140435161722805153518  1766.617637512327416056724737722775   
166  1883644472854031162817139514072676  1769.136310824577376846199888707781   

                                 token1Price    tick  \
0    0.0005787577299408856894168926824109645  201774   
1    0.0005764559547495751999689070796788115  201734   
2    0.0005728123852958934694297346816470754  201670   
3    0.0005733154834167925578956548047763606  201679   
4    0.0005674984316750099366809510866950609  201577   
..                                       ...     ...   
162  0.0005693679402173520241956610360252131  201610   
163  0.0005633798221277398304558339587341144  201504   
164  0.0005653912219036140951601194258979986  201540   
165  0.0005660534451632417961020896335622054  201552   
166  0.0005652475696086468553592753961406267  201537   

                   feeGrowthGlobal0X128  \
0    2312158785098937052494570152217525   
1    2312216559855034939422879201028187   
2    2312311586355258922806996249821275   
3    2312405216177388689787126610454526   
4    2312562964065321587002564097838970   
..                                  ...   
162  2325689226524357531016841471727450   
163  2325808377051363646724557867581639   
164  2326008139763432932331735673541946   
165  2326136592649447151094714434389576   
166  2326305393609504734774243429534734   

                            feeGrowthGlobal1X128  \
0    1109665930631382897574108289654189272954102   
1    1109691771721878689625682223884054944269262   
2    1109733988873390402439915596012262413890369   
3    1109788781038653813021677282971350131285545   
4    1109857533797043930382095487494978532388488   
..                                           ...   
162  1117182664696319171891220060386964137427102   
163  1117228559926179730824551731671866657297461   
164  1117348097322167256295674311880980576086263   
165  1117422968088234277235903572926793760245590   
166  1117515847583285328747004696134020800288134   

                                  tvlUSD volumeToken0 volumeToken1 volumeUSD  \
0    363741182.8303982633120492904007652            0            0         0   
1    364112711.2526927733521560182183211            0            0         0   
2    361535632.0601155211202920489855564            0            0         0   
3    364444641.3551118523700690778229295            0            0         0   
4    365361117.7910602923638770894625618            0            0         0   
..                                   ...          ...          ...       ...   
162  373281051.1590490239509933928722079            0            0         0   
163  382032389.515594492

In [51]:
my_ratio = 286206700622530/30788832203141306590 # 暂时随便选了个数算了个大概

swaps1["fee0"] = abs(swaps1["amount0"] * my_ratio) * 0.0005
fee0_sum = swaps1["fee0"].sum()

swaps1["fee1"] = abs(swaps1["amount1"] * my_ratio) * 0.0005
fee1_sum = swaps1["fee1"].sum()

print(fee0_sum)
print(fee1_sum)
print('{:.0f}'.format(fee0_sum))
print('{:.0f}'.format(fee1_sum))
# 对比这个比例和revert.finance上的，相差不大，算的应该是对的，但是数值对不上，我发现可能是swaps数据没有拿全，用query_swaps拿到的每次都只有100条
# 解决： 这是graph ql的特征，每次只返回100行。需要用分页查询来查询所有的数据

28.365540468724014
0.015938399029110484
28
0


In [52]:
print('this is sqrt price lower: {:.0f}'.format(sqrt_p_lower))
print('this is sqrt price upper:{:.0f}'.format(sqrt_p_upper))

sqrt_p_lower_f = '{:.0f}'.format(sqrt_p_lower)
sqrt_p_upper_f = '{:.0f}'.format(sqrt_p_upper)

this is sqrt price lower: 1639390289596216154630867994542080
this is sqrt price upper:2138941539491792332891588793466880


In [53]:
begin = datetime(2023, 3, 19)
end = datetime.now()
# 筛选出所有在头寸内的swaps
data = query.query_swaps(begin=int(begin.timestamp()), end=int(end.timestamp()), pool_id=pool_id, sqrtPriceX96_gte=sqrt_p_lower_f, sqrtPriceX96_lte=sqrt_p_upper_f)

this is the cursor: 0x02a679e24a73775694639ae501a83ce9dda883369f01a6e0ab8083d43f27f80a#4711857
this is the cursor: 0x0558dac2cef364cc3e2e19071953e1ee0982d401fd0189692e4854451c3c1109#4681779
this is the cursor: 0x0830d47b8f13c2eef7896d35d43b831cc934c0256ed85d05b947b9525700cc2f#4682686
this is the cursor: 0x0addc615a9fc668efd97ad3eca1c5c1a916faeb66ca2b1a68f1f2d5f97c5eb92#4636808
this is the cursor: 0x0da9aa1424a812801a5ba13b0e9629f625edfea5c3bea07150d5d83a03baa5f0#4704420
this is the cursor: 0x1097fd6b9a0b63e8acba28d890c29d4556c850e232ef769aeee2bebcd95d6b96#4630319
this is the cursor: 0x135de3964f4ed1b95440de1b9d43e8a88514383822af49028a3c5f7c21b27e8c#4622918
this is the cursor: 0x162526172839deda80cfcb566e73983b43398bd9ded3af7e19884277378cb61b#4705329
this is the cursor: 0x18eb76bad49b905763af84daf579038c202fb2883e14f8d4c92e8c3abe9f9ffa#4680299
this is the cursor: 0x1bbadede6a7d7a3b420336f844a986967a3ebcbcab30fabcea1e1f27537107e4#4688651
this is the cursor: 0x1e6abab721b8507830d3d35ce8ba

In [58]:
data.shape[0]

92338

In [59]:
data.head()

id   timestamp  \
0  0x00005be3feffa40600371196f6ebe99484cef2e030cb...  1679680919   
1  0x00005e650a5a29a7340961ee1070bcb4d699a5c6b47c...  1680218951   
2  0x00010099f534a61a99fa906e57cabd736e9f7e43ee0f...  1679774651   
3  0x000128f2e478320e005d9435b14ad71341d06b8d1d7f...  1679404235   
4  0x0001770e8cba1be004b066570f49bf727e1b00774d32...  1680084887   

                                       sender  \
0  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
3  0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45   
4  0x7122db0ebe4eb9b434a9f2ffe6760bc03bfbd0e0   

                                    recipient  \
0  0x697d1eeb266aec5251901e20e6a09dbdb68008c1   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xd22125cb225f39fe6aa0c16afe6b75872a3529d1   
3  0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5   
4  0x1111111254eeb25477b68fb85ed929f73a960582   

                                       origin        amount0  \
0  0x697d1eeb266aec5251901e20e6a09dbdb68008c1  -19922.526611   
1  0xc0c441bd9889f30fd219478a985e84fac8c2a706    1104.872406   
2  0xd22125cb225f39fe6aa0c16afe6b75872a3529d1  -12033.180287   
3  0xfa7d0cfee4d881cf8159e0d18c692ccb71e9c0c5   -5000.197259   
4  0xc6c7565644ea1893ad29182f7b6961aab7edfed0    -794.907298   

                 amount1                            amountUSD  \
0  11.442382101497925129  19903.00432076901806488783284947735   
1  -0.617085078176457911  1103.966933243716954948609702966833   
2                      7  12040.45320764383144992175201712345   
3   2.750357643401229741  4998.420143808785696912599946213698   
4   0.438164882921890842   795.916771734336092480722470468768   

                         sqrtPriceX96    tick logIndex  
0  1898284365247915042890667106569354  201692      295  
1  1872856442624146468221690277382731  201422      369  
2  1910435929075656551647679638334639  201820      272  
3  1857690380746107826980693568912245  201260      209  
4  1859651863181264024198474611348102  201281       11

In [60]:
# 现在要做的就是把swaps表格和liquidity表格合并！
# 首先我们已经有了池子liquidity所有tick的流动性数据，需要计算出池子里在我们的头寸范围内的总流动性
# 我本来以为算这个会很麻烦，但是我发现query_liquidity函数返回的数据中，liquidity就是in range的liquidity。
liquidity_data = query.query_liquidity(begin=int(begin.timestamp()), end=int(end.timestamp()), pool_id=pool_id, sqrtPrice_gt=sqrt_p_lower_f, sqrtPrice_lt=sqrt_p_upper_f)

this is the cursor: 1680436800


In [61]:
liquidity_data = liquidity_data[["periodStartUnix", "liquidity"]]
liquidity_data = liquidity_data.rename(columns={"periodStartUnix": "timestamp", "liquidity": "liquidity"})

In [62]:
# 先排序
data["timestamp"] = data["timestamp"].astype(int)
data = data.sort_values(by="timestamp")
liquidity_data = liquidity_data.sort_values(by="timestamp")
# 找liquidity data只时间最接近的来合并
merged_data = pd.merge_asof(
    data,
    liquidity_data[["timestamp", "liquidity"]],
    on="timestamp",
    direction="nearest"
)

In [63]:
merged_data.head(3)

id   timestamp  \
0  0xdb0dc52d01c31207494402aadc51386666b6bc7a5723...  1679155211   
1  0xf20cd95e06fae029ea9d9b496ec03b53488e905796ee...  1679155223   
2  0x0861d00acafcd74202e82741c031db7417fbca4c45fe...  1679155223   

                                       sender  \
0  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   

                                    recipient  \
0  0x3d889b6391506032f9e9380a8f5cd3199661fb72   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xe63df014b8481ecc14582ff09e88809b0ad4e5a9   

                                       origin       amount0  \
0  0x3d889b6391506032f9e9380a8f5cd3199661fb72   -7168.36401   
1  0xfb8c0777051218d3ecb6ade85783b64ce54ad409   1792.878253   
2  0xe63df014b8481ecc14582ff09e88809b0ad4e5a9  -17615.76052   

                amount1                            amountUSD  \
0    4.0021259669077627  7167.505227908871519372434303813185   
1                    -1  1791.794058488190625795139705811723   
2  9.835103915662891668  17613.78905750513568184175674883613   

                         sqrtPriceX96    tick logIndex             liquidity  
0  1871575388396658563917411167894519  201409      498  31691854789759819627  
1  1871597848374746597023794122478978  201409      414  31691854789759819627  
2  1871596376831572468181234428468238  201409       17  31691854789759819627

In [64]:
my_position = query.query_position(position_id=position_id)
position_liquidity = my_position["liquidity"]
merged_data['liquidity'] = merged_data['liquidity'].astype(float)
merged_data['ratio'] = float(position_liquidity) / merged_data["liquidity"]

In [65]:
merged_data.head(3)

id   timestamp  \
0  0xdb0dc52d01c31207494402aadc51386666b6bc7a5723...  1679155211   
1  0xf20cd95e06fae029ea9d9b496ec03b53488e905796ee...  1679155223   
2  0x0861d00acafcd74202e82741c031db7417fbca4c45fe...  1679155223   

                                       sender  \
0  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   

                                    recipient  \
0  0x3d889b6391506032f9e9380a8f5cd3199661fb72   
1  0xef1c6e67703c7bd7107eed8303fbe6ec2554bf6b   
2  0xe63df014b8481ecc14582ff09e88809b0ad4e5a9   

                                       origin       amount0  \
0  0x3d889b6391506032f9e9380a8f5cd3199661fb72   -7168.36401   
1  0xfb8c0777051218d3ecb6ade85783b64ce54ad409   1792.878253   
2  0xe63df014b8481ecc14582ff09e88809b0ad4e5a9  -17615.76052   

                amount1                            amountUSD  \
0    4.0021259669077627  7167.505227908871519372434303813185   
1                    -1  1791.794058488190625795139705811723   
2  9.835103915662891668  17613.78905750513568184175674883613   

                         sqrtPriceX96    tick logIndex     liquidity     ratio  
0  1871575388396658563917411167894519  201409      498  3.169185e+19  0.000009  
1  1871597848374746597023794122478978  201409      414  3.169185e+19  0.000009  
2  1871596376831572468181234428468238  201409       17  3.169185e+19  0.000009

In [66]:
merged_data['amount0'] = merged_data['amount0'].astype(float).apply(lambda x: max(0, x))
merged_data['amount1'] = merged_data['amount1'].astype(float).apply(lambda x: max(0, x))

merged_data["fee0"] = merged_data["amount0"] * merged_data['ratio'] * 0.0005
fee0_sum = merged_data["fee0"].sum()

merged_data["fee1"] = merged_data["amount1"] * merged_data['ratio'] * 0.0005
fee1_sum = merged_data["fee1"].sum()

print(fee0_sum)
print(fee1_sum)

17.033364499275496
0.009627564829887168
